plugin/audit_log/audit_log.c

```diff
@@ -411,40 +450,51 @@ size_t audit_log_general_record(char *buf, size_t buflen,
 
-  return my_snprintf(buf, buflen,
-                     format_string[audit_log_format],
-                     name,
-                     make_record_id(id_str, sizeof(id_str)),
-                     make_timestamp(timestamp, sizeof(timestamp), t),
-                     event->general_sql_command.str,
-                     event->general_thread_id,
-                     status,
-                     escape_string(event->general_query,
-                                   event->general_query_length,
-                                   query, sizeof(query), NULL),
-                     escape_string(event->general_user,
-                                   event->general_user_length,
-                                   endptr, endtmp - endptr, &endptr),
-                     escape_string(event->general_host.str,
-                                   event->general_host.length,
-                                   endptr, endtmp - endptr, &endptr),
-                     escape_string(event->general_external_user.str,
-                                   event->general_external_user.length,
-                                   endptr, endtmp - endptr, &endptr),
-                     escape_string(event->general_ip.str,
-                                   event->general_ip.length,
-                                   endptr, endtmp - endptr, &endptr));
+  query= escape_string(event->general_query, event->general_query_length,
+                       endptr, endbuf - endptr, &endptr, &full_outlen);
+  user= escape_string(event->general_user, event->general_user_length,
+                      endptr, endbuf - endptr, &endptr, &full_outlen);
+  host= escape_string(event->general_host.str, event->general_host.length,
+                      endptr, endbuf - endptr, &endptr, &full_outlen);
+  external_user= escape_string(event->general_external_user.str,
+                               event->general_external_user.length,
+                               endptr, endbuf - endptr, &endptr, &full_outlen);
+  ip= escape_string(event->general_ip.str, event->general_ip.length,
+                    endptr, endbuf - endptr, &endptr, &full_outlen);
+
+  buflen_estimated= full_outlen * 2 +
+                    strlen(format_string[audit_log_format]) +
+                    strlen(name) +
+                    strlen(event->general_sql_command.str) +
+                    20 + /* general_thread_id */
+                    20 + /* status */
+                    MAX_RECORD_ID_SIZE + MAX_TIMESTAMP_SIZE;
+  if (buflen_estimated > buflen)
+    return buflen_estimated;
+
+  len= my_snprintf(endptr, endbuf - endptr,
+                   format_string[audit_log_format],
+                   name,
+                   make_record_id(id_str, sizeof(id_str)),
+                   make_timestamp(timestamp, sizeof(timestamp), t),
+                   event->general_sql_command.str,
+                   event->general_thread_id,
+                   status, query, user, host, external_user,ip);
```

Comment:
Can we end up in a situation where buffer was enough to escape all the variables above, but not enough to snprintf without truncation?